### DISCUSSION TITLE
Embedding native surface while still using Avalonia as host

### DISCUSSION THREAD

**Discussion by jp2masa** [AvaloniaUI/Avalonia#9268] [Q&A]
## Use case

Veldrid doesn't seem to like OpenGL, so `OpenGlControlBase` can't be used. So I changed to a `NativeControlHost` where I create a SDL2 window from native API and pass it to Veldrid, which works just fine, but I'd like to use an Avalonia Window and not from SDL2, to make it easier to handle inputs and have less dependencies.

Is this possible?

**Comment by maxkatz6** [comment]
I think `var window = new EmbeddableControlRoot()` is what you need.


See:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Embedding/EmbeddableControlRoot.cs#L18
And https://github.com/AvaloniaUI/Avalonia/blob/8b5f97e919252aabc1964fb2e0a3efece67413d2/src/Windows/Avalonia.Win32/EmbeddedWindowImpl.cs#L11

**Reply by jp2masa** [reply]
Unfortunately it doesn't work... a few problems:

1. API: `CreateWindowOverride` is called by the base constructor, so I have no way to pass it the parent handle (except hacky static variable, which I'm using for now)
2. `PointerPressed` doesn't work, also tried to set `Background = Brushes.Transparent` (and White)
3. Visual glitches (this may be related to Vulkan setup, still have to check)

**Reply by jp2masa** [reply]
It looks like the visual glitches were caused by not having `app.manifest`. When I disabled the compositor I got the expected exception (#8609), but the compositor was actually allowing me to run the app and it was totally broken.

Problems 1. and 2. remain, although I checked now and `KeyDown` works, but `PointerPressed` and `PointerWheelChanged` don't.

**Reply by maxkatz6** [reply]
But you can [set a parent](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setparent) to the window after it was created, can't you?
Not sure what's causing second problem.

**Reply by jp2masa** [reply]
That makes sense, although it's a lot more work (had to update style to WS_CHILD after that as well, can't be done on CreateWindowEx because there's no parent yet).

Still no idea about pointer events... Also, is there a more cross platform way of doing this? Anything planned?

Thanks for all the help!

**EDIT:** also tried `AddHandler(PointerPressedEvent, (sender, e) => OnPointerPressed(e), handledEventsToo: true);` and it doesn't work...

**Reply by maxkatz6** [reply]
> Also, is there a more cross platform way of doing this? Anything planned?

Depends on what do you mean here.
`var window = new EmbeddableControlRoot()` is a cross platform way to create a window that can be embedded.

> Still no idea about pointer events... 

Where are you trying to handle pointer events? On which control?
If you are handling it on the NativeControlHost, then I guess it's not a surprise. Since child window might handle all the events. Related:
https://github.com/AvaloniaUI/Avalonia/issues/5256
https://github.com/AvaloniaUI/Avalonia/issues/8104

**Reply by jp2masa** [reply]
> Depends on what do you mean here.
var window = new EmbeddableControlRoot() is a cross platform way to create a window that can be embedded.

Yes, but I still have to create a custom `WindowImpl` for each platform.

> Where are you trying to handle pointer events? On which control?
If you are handling it on the NativeControlHost, then I guess it's not a surprise. Since child window might handle all the events. Related:
https://github.com/AvaloniaUI/Avalonia/issues/5256
https://github.com/AvaloniaUI/Avalonia/issues/8104

I know, I already read those, I'm trying to handle the events on the embeddable control root, it should have its own wnd proc, right?

**Reply by maxkatz6** [reply]
> Yes, but I still have to create a custom WindowImpl for each platform.

Do you create custom window impl just to set a parent? Just for case NativeControlHost automatically sets it from what I remember.

> it should have its own wnd proc, right?

Right

**Reply by jp2masa** [reply]
> Do you create custom window impl just to set a parent? Just for case NativeControlHost automatically sets it from what I remember.

I just tried to use the default window impl (top level impl has no handle so I still had to create `new Avalonia.Win32.WindowImpl()`), and it has 2 problems: parent is not set automatically and it doesn't have WM_CHILD attribute, so it really has to be a custom window impl I guess...

**Reply by maxkatz6** [reply]
You can also use `NativeControlHost.CreateDefaultChild()` it won't return full window impl, but can be used to host content in the NativeControlHost itself. For reference, that's Win implementation with WM_CHILD https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32NativeControlHost.cs#L84 and parent.

**Reply by jp2masa** [reply]
It's actually `INativeControlHostImpl.CreateDefaultChild()`, and I think that doesn't help much, because it returns a handle to a new native window, and I would need an impl object to pass to embedded control root.

**Reply by maxkatz6** [reply]
This method is available on the [NativeControlHost too](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/NativeControlHost.cs#L164-L169) (different name though, haven't noticed before - `base.CreateNativeControlCore()`), you don't need implementation.
> because it returns a handle to a new native window

But doesn't it work as a replacement for SDL2 window? You can pass this handle to the Veldrid from what I understand.

**Reply by jp2masa** [reply]
Yes, but I'm no longer using SDL2, I'm trying to use Avalonia only, and passing the handle to Veldrid for rendering. Unfortunately the embedded control root exposes the platform impl as a top level impl (makes sense considering not all platforms use windows for embedding), but top level impl doesn't give me a handle.

The main problem is that pointer input doesn't work.

**Reply by anose001** [reply]
@jp2masa, did this ever work out for you?
